Added support for parsing mumble messages

timokoesters opened the pull request:
A connection object now has a read method which return a Result<MessageType, _>. A MessageType is an enum with every mumble message that is in the specification. Most of them have a protobuf/mumble Message struct associated with them except for UDPTunnel, because it's never used and a few Box<> types with the data in it (Because clippy complained that some variants were too large). I also implemented a users vector, which stores the users and updates their channel attribute when they switch between channels (It doesn't delete them when they disconnect yet).

What I like about the way of reading messages is that you can match the type of the message which looks quite nice (see example).

For sending messages there are a few methods in the connection struct ( send_text_message, send_authentication, ... ) which allow sending messages to the mumble server. They are all based on the send method which takes a message and converts it to a byte array which is send to the server. A problem is that one needs the type/id of the message in order to send it. I solved this by implementing a HasType trait for every message, which works but is quite ugly in my opinion.

Please tell me what you think about this approach.

Prior99 commented:
@ahti would you have a look?

Prior99 commented:
I generally like the approach, however I added some friends as reviewers that have a better understanding of Rust than I have.

Is it the correct approach to add the code generated from protobuf into the repository? Shouldn't a `build.rs` handle that?

Prior99 commented on src/connection.rs:
```diff
@@ -0,0 +1,205 @@
+        ctx.set_verify(SslVerifyMode::NONE);
+        ctx.set_certificate_chain_file(cert).unwrap();
+        ctx.set_private_key_file(&key, SslFiletype::PEM).unwrap();
+        ctx.set_cipher_list("ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:AES128-G ...
```
Should this be hardcoded and non-overridable? Does Mumble enforce this set of ciphers in it's standard?

timokoesters replied:
I left this method like it was before, but according to [this](https://mumble-protocol.readthedocs.io/en/latest/overview.html#f1) "The TCP control channel uses TLSv1 AES256-SHA".

Prior99 commented on src/connection.rs:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    pub fn send<M: Message + HasType>(&mut self, message: &M) {
+        // Create an stream to ouput this message
```
I don't require it, but I would appreciate it if you could correct obvious spelling mistakes like these.

Prior99 commented on src/connection.rs:
```diff
@@ -0,0 +1,205 @@
+        let mut result = Vec::new();
+
+        // Append id
+        result.write_u16::<BigEndian>(message.get_id()).unwrap();
```
I think we shouldn't use `unwrap` here, this can fail (if the connection was lost).

timokoesters replied:
I changed it to an expect, but is this the preferred way or how should we do this?

Prior99 commented on src/connection.rs:
```diff
@@ -0,0 +1,205 @@
+        // Append payload
+        result.append(&mut payload);
+
+        //println!("Sending bytes: {:x?}", result);
```
Please remove this line.

Prior99 commented on src/connection.rs:
```diff
@@ -0,0 +1,205 @@
+            message.set_channel_id(channel_id);
+            self.send(&message)
+        } else {
+            println!("Connection not established. (No session)");
```
Instead of using `println!`, please add a `log` crate and use the `debug!` macro. Or at least open an issue for that.

Prior99 commented on src/message_types.rs:
```diff
@@ -0,0 +1,249 @@
+    }
+}
+
+impl HasType for mumble::SuggestConfig {
```
The content of this file looks hard to maintain. Is there no better way to do this? @svenstaro

svenstaro replied:
This doesn't seem like a good solution to me either.

timokoesters replied:
Can you take a look again? I got rid of the HasType enum and `match`ed them instead. This is already much better.

svenstaro commented on src/connection.rs:
```diff
@@ -0,0 +1,203 @@
+        output_stream.flush().unwrap();
+    }
+
+    pub fn send_version(&mut self) {
```
All functions (and especially pub functions) should have docstrings.

svenstaro commented:
Is it really necessary to check in the generated code? Can't this be generated in build.rs?

ahti commented on src/message_types.rs:
```diff
@@ -0,0 +1,141 @@
+    Reject(mumble::Reject),
+    ServerSync(mumble::ServerSync),
+    ChannelRemove(mumble::ChannelRemove),
+    ChannelState(Box<mumble::ChannelState>),
```
Haven't looked into this further, but please make sure to only Box where required (`ChannelState`, `UserState` and `UserStats`). If it is required here, please add a comment explaining why.

timokoesters replied:
According to clippy, some of the structs were much larger than other ones and it recommended to put them into Boxes so they don't make every enum item bigger.

timokoesters replied:
Would it be better / more readable if we put every item in a Box?

ahti commented on src/connection.rs:
```diff
@@ -0,0 +1,188 @@
+    }
+}
+
+pub fn connect(url: &str, key: &Path, cert: &Path, username: &str, password: &str) -> Connection {
```
The password here (and in `send_authentication`) should be an optional I think (https://mumble-protocol.readthedocs.io/en/latest/establishing_connection.html#authenticate).

timokoesters replied:
Should I just take an Option type as an argument or create two seperate constructors?

ahti replied:
An Option sounds good 👍

ahti commented on src/connection.rs:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    pub fn keep_alive(&mut self) {
```
I don't think I like the polling approach here. E.g. I think `examples/connect.rs` would get disconnected if there is no valid message from the server for 30s. The ping sounds like something the connection should handle internally, at least by default.

Thoughts? @Prior99 @svenstaro 

Either way, `SystemTime` is not monotonic and not the right tool here. `Instant` should be better.

timokoesters replied:
I switched to `Instant`, but we still have polling.

ahti commented on src/connection.rs:
```diff
@@ -0,0 +1,188 @@
+/// A way to interact with the Mumble server
+pub struct Connection {
+    stream: SslStream<TcpStream>,
+    session: Option<u32>,
```
Since most real functionality only works once a session is established: Could we split this into two types, one for a connection that is in the process of being initialized (which should probably not even be exposed to the user using the default API), and one for a connection that had it's session initialized and is ready to send text messages, change channel, etc?

Edit: Thinking about this some more, the connection should probably loose the session, user list, etc. and just be a way of sending and receiving messages. A separate `Session` struct that wraps a connection should hold session id, user lists, etc.

timokoesters replied:
This seems like a good idea. I'll make a new session struct.

timokoesters replied:
Ok, I just committed the change. Can you take a look?

ahti commented on src/connection.rs:
```diff
@@ -0,0 +1,188 @@
+        Arc::new(ctx.build())
+    }
+
+    pub fn read(&mut self) -> Result<MessageType, std::io::Error> {
```
This implementation seems very fragile. `ssl_read` does not necessarily block to fill the whole buffer, but returns how many bytes were actually read. The rest will be left as zeroes.

timokoesters replied:
I couldn't find an ssl_read_exact method. What would you recommend to do?

timokoesters replied:
I switched to read_exact which seems to work.

timokoesters commented:
Thanks for your review. I'll try to do my best to follow your suggestions.

svenstaro commented on examples/connect.rs:
```diff
@@ -1,24 +1,24 @@
 extern crate rumble;
```
Can't you remove this on 2018 edition?

timokoesters replied:
Oops, forgot that. Yes that's possible.

svenstaro commented:
This remains unanswered: https://github.com/Prior99/rumble/pull/2#issuecomment-442924082

timokoesters commented:
It might be more useful to have it in the repo so that one doesn't need to have `protoc` installed to compile it.

svenstaro commented:
I think depending on protobuf being installed is fine. It's more practical and less error-prone than putting precompiled stuff here. All distros ship protobuf anyway.

timokoesters commented:
I found the [protoc-rust](https://github.com/stepancheg/rust-protobuf/tree/master/protoc-rust) crate and added it as a build dependency. One still needs to have protoc installed but not rust plugins. I added some code to the build.rs file to generate the mumble.rs file automatically. (I also added src/mumble.rs to the .gitignore file)

Prior99 commented:
Thanks for your contribution and staying with us through this. Do you want to be a contributer or own this project?

timokoesters commented:
Thanks for merging! I'd like to be a contributor for now. This is my first major contribution to an open source project and I'm happy it worked out.

timokoesters commented:
I think the next big step for this project should be UDP communication so one can play music from a local file or stream some (using youtube-dl)

Prior99 commented:
I added you as an collaborator.
